Sources/Datadog/Tracing/Utils/URLSessionSwizzler.swift

```diff
@@ -6,102 +6,157 @@
+                            blockTask?.payload?(.completed)
+                        }
+                        let task = currentTypedImp(impSelf, Self.selector, interceptionResult.modifiedRequest, modifiedCompletion)
+                        try? Resume.swizzleIfNeeded(in: task)
+                        task.payload = interceptionResult.taskObserver
+                        blockTask = task
+                        return task
+                    }
+                    return currentTypedImp(impSelf, Self.selector, impURLRequest, impCompletion)
+                }
+            }
         }
-        let newImp: IMP = imp_implementationWithBlock(newImpBlock)
-        try swizzler.swizzle(selector: sel, in: subjectClass, with: newImp)
     }
 
-    static func swizzleDataTaskWithRequestCompletionHandler() throws {
-        typealias TypedIMP = TypedIMPs.DataTaskWithURLRequestCompletion
-        typealias TypedBlockIMP = TypedBlocks.DataTaskWithRequestCompletion
-        let sel = Selectors.DataTaskWithRequestCompletion
+    private enum Resume {
+        private static let selector = #selector(URLSessionTask.resume)
+        private typealias TypedIMP = @convention(c) (URLSessionTask, Selector) -> Void
+        private typealias TypedBlockIMP = @convention(block) (URLSessionTask) -> Void
+
+        /// NOTE: RUMM-452
+        /// URLSessionTask.resume is not called by its subclasses!
+        /// Therefore, we swizzle this method in the subclass.
+        /// This is unlike swizzling dataTaskURL/Request: in URLSession base class
+        static func swizzleIfNeeded(in task: URLSessionTask) throws {
+            guard let taskClass = object_getClass(task) else {
+                userLogger.debug("\(task) does not have a class!")
+                return
+            }
+            let foundMethod = try swizzler.findMethod(with: selector, in: taskClass)
+            // NOTE: RUMM-452 We will probably not need to swizzle tasks every time
+            // "onlyIfNonSwizzled: true" lets us perform swizzling in given class if not done before
+            swizzler.swizzle(
+                foundMethod,
+                impSignature: TypedIMP.self,
+                impProvider: { currentTypedImp -> TypedBlockIMP in
+                    return { impSelf in
+                        impSelf.payload?(.starting)
+                        return currentTypedImp(impSelf, selector)
+                    }
+                },
+                onlyIfNonSwizzled: true
+            )
+        }
+    }
+}
 
-        let typedOriginalImp: TypedIMP = try swizzler.currentImplementation(of: sel, in: subjectClass)
+private extension URLSession {
+    /// This method is used in swizzled dataTaskWithURL implementation
+    /// We create URLRequest from URL and pass it to interceptor
+    func urlRequest(with url: URL) -> URLRequest {
+        return URLRequest(url: url, cachePolicy: configuration.requestCachePolicy, timeoutInterval: configuration.timeoutIntervalForRequest)
```

Comment:
What about starting with the safest possible solution:
* 1/ `session.dataTask(with: request, completion: {})` → send span + propagate trace
* 2/ `session.dataTask(with: url, completion: {})` → send span

With extension to:
* 3/ `session.dataTask(with: request)` → redirect to 1, so: send span + propagate trace
* 4/ `session.dataTask(with: url)` → redirect to 2, so: send span

This is the most safe & valuable implementation I can come up with. WDYT @buranmert ?

Reply:
methods without completion handler (3&4) use `session.delegate` for events, changing their execution path will break their behavior, so i'd say let's leave them out of this.

what would be confusing, distributed tracing for requests but not for urls.
some SDKs provide `DistributedTracing` as a configurable feature.

if we add `DistributedTracing` to `TracingConfiguration`, 
then we can provide the same behavior (span + trace) for both methods 
with an emergency brake ("if it breaks anything, disable `DistributedTracing`")
wdyt @ncreated ?

Reply:
> methods without completion handler (3&4) use session.delegate for events, changing their execution path will break their behavior, so i'd say let's leave them out of this.

Ah I see, totally right 👌.

> if we add DistributedTracing to TracingConfiguration,
then we can provide the same behavior (span + trace) for both methods
with an emergency brake ("if it breaks anything, disable DistributedTracing")
wdyt @ncreated ?

As concluded on Slack with product - let's start simple with
- `session.dataTask(with: request, completion: {})` → send span + propagate trace
- `session.dataTask(with: url, completion: {})` → send span

and we'll consider the propagation for plain `URLs` later, if requested. This requires a round of simplification in `URLSessionSwizzler` and its tests 🚀.

Reply:
1. `dataTaskWithURL` doesn't redirect to `dataTaskWithRequest` anymore
2. `RequestInterceptor` changed to `(URL?) -> (TaskObserver, HTTPHeaders)?`
    * so that we still have only one way to intercept requests

pushed ✅